h5py/h5t.pyx

```diff
@@ -1994,3 +2004,80 @@ cpdef object py_get_vlen(object dt_in):
     warn("Deprecated; use check_vlen_dtype(dtype) instead",
         H5pyDeprecationWarning)
     return check_vlen_dtype(dt_in)
+
+
+def register_dtype(dtype dt_in, bytes tag=None):
+    """ (dtype dt_in, bytes tag=None)
+
+    Register a NumPy dtype for use with h5py. Types registered in this way
+    will be stored as a custom opaque type, with a special tag to map it to
+    the corresponding NumPy type.
+
+    Opaque types with this tag will be mapped to NumPy types in the same way.
+
+    The default tag is generated via the code:
+    ``b"NUMPY:" + dt_in.descr[0][1].encode()``.
+    """
+    try:
+        if not np.issubdtype(dt_in, np.void):
+            py_create(dt_in)
+        raise RuntimeError("This tag or dtype is already registered or native.")
+    except TypeError:
+        pass
+
+    if tag is None:
+        tag = b"NUMPY:" + dt_in.descr[0][1].encode()
+
+    cdef TypeID new_type = typewrap(H5Tcreate(H5T_OPAQUE, dt_in.itemsize))
+    new_type.set_tag(tag)
+
+    _dtype_map[dt_in] = new_type
+    _tag_map[tag] = dt_in
+
+
+def deregister_dtype(object obj):
+    """ (dtype obj OR bytes obj)
+
+    Deregister a dtype/tag from the NumPy-tag mapping, along with the.
+    corresponding tag/dtype.
+    """
+    if isinstance(obj, dtype):
+        _deregister_dtype(obj)
+    elif isinstance(obj, bytes):
+        _deregister_tag(obj)
+    else:
+        raise TypeError("obj must be either a dtype or bytes object")
+
+
+def _deregister_dtype(dtype dt_in):
+    """ (dtype dt_in)
+
+    Deregister a dtype from the NumPy-tag mapping, along with the.
+    corresponding tag.
+    """
+    if dt_in not in _dtype_map:
+        raise RuntimeError("This dtype is not registered.")
+
+    cdef TypeID tid_hl = _dtype_map[dt_in]
+    cdef bytes tag = tid_hl.get_tag()
+    cdef hid_t tid = tid_hl.id
+    del _dtype_map[dt_in]
+    del _tag_map[tag]
+    H5Tclose(tid)
+
+
+def _deregister_tag(bytes tag):
+    """ (bytes tag)
+
+    Deregister a tag from the NumPy-tag mapping, along with the.
+    corresponding dtype.
+    """
+    if tag not in _tag_map:
+        raise RuntimeError("This tag is not registered.")
```

Comment:
I wonder if this and the corresponding 'dtype not registered' error should be KeyError, rather than the very generic RuntimeError. I haven't thought much about the exception semantics, though.

Reply:
From the [docs](https://docs.python.org/3/library/exceptions.html#KeyError):

> Raised when a mapping (dictionary) key is not found in the set of existing keys.

Here, it would be confusing as the user doesn't see the map directly. How about `ValueError`?